### PR TITLE
Fix StackOverflow with equals check of maps with cyclic references

### LINKED ISSUE #33929: Make == deal with cycles
Following example should work. Currently throws `StackOverflow`.
```ballerina
import ballerina/io;

type Part anydata[];
type J anydata;

public function main() {
    Part yin = [];
    Part yang = [];
    yin[0] = yang;
    yang[0] = yin;
    io:println(yin == yang); // @output true

    Part father = [];
    Part son = [];
    Part sprit = [];
    father[0] = son;
    son[0] = sprit;
    sprit[0] = father;
    io:println(father == son); // @output true

    io:println(father == yin); // @output true

    map<J> j1 = { loop: () };
    map<J> j2 = { loop: () };
    j1["loop"] = j1;
    j2["loop"] = j1;
    map<J> j3 = { loop: () };
    j3["loop"] = { loop: { loop: { loop: j3 }}};
    io:println(j1 == j3); // @output true
}
```

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #33929

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValuePair.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    Set<Object> valuePairSet = new HashSet<>(2);
+
+    public ValuePair(Object obj1, Object obj2) {
+        valuePairSet.add(obj1);
+        valuePairSet.add(obj2);
+    }
+
+    public boolean equals(Object o) {
+        if (!(o instanceof ValuePair valuePair)) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValuePair.java:40]: Do we need this code?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/AbstractArrayValue.java [hunk 2]
```diff
@@ -73,6 +75,29 @@ public void append(Object value) {
...
+    @Override
+    public boolean equals(Object o, Set<ValuePair> visitedValues) {
+        ValuePair compValuePair = new ValuePair(this, o);
+        for (ValuePair valuePair : visitedValues) {
+            if (valuePair.equals(compValuePair)) {
+                return true;
+            }
+        }
+        visitedValues.add(compValuePair);
+
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/AbstractArrayValue.java:87]: Shall we remove unwanted new lines

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValuePair.java (hunk 1 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ValuePair.java:40]: It seems, there are no possibilities to this check being executed. Btw, IDEA gives warning and suggests this check. I think we can replace it with casting directly.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/AbstractArrayValue.java (hunk 2 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/AbstractArrayValue.java:87]: I added it to improve the readability as following lines belongs to a separate logic. Isn't it okay to keep?

**Review by HindujaB — APPROVED** [review]
LGTM

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**